### DISCUSSION TITLE
Джойстик для Н4

### DISCUSSION THREAD

**Discussion by Gescha5555** [kachurovskiy/nanoels#185] [Q&A]
https://youtu.be/SPQWSbeQXGE

Приветствую всех. 
Месяц назад собрал я ELS H4 по проекту Максима но подружиться с её управлением у меня терпения не хватило.
Решил я добавить джойстик , показываю что я пока сделал .
Полный функционал пока в асинхронном режиме в остальных частично.
Ещё попытаюсь залезть в некоторые режимы но там много переписывать придётся так как Максим
изначально писал под свою логику управления .
Удачи всем в творчестве.

**Comment by vegetate7** [comment]
Круто. Еще-бы что-б само выключалось на стопах. Лежит такой переключатель, все посматриваю на него, как-бы сделать электромагнитное удержание.

**Comment by Gescha5555** [comment]
Извините немного не понял , на электронных стопах он останавливается .
И электромагнитное удержание куда хотите подключить?

**Reply by vegetate7** [reply]
Не останавливалось, а чтобы джойстик в среднее положение возвращался. Видел самодельные такие выключатели на ютубе. Выключатель без фиксации, удерживается электромагнитом. выключается либо рукой, либо отключением электромагнита.

**Reply by Gescha5555** [reply]
Я вас понял про джойстик с автоматическим отключением ,я не понял зачем ему отключаться в нейтраль если станок доехав до 
установленного электронного стопа останавливается  , чтобы поехать на зад надо перевести джойстик в другую сторону и
ему всё равно с магнитной фиксацией или с обычной всё равно включать руками.

**Reply by vegetate7** [reply]
В плане дополнительного функционала - совсем незачем. Исключительно для эстетического соответствия режима работы станка и положения джойстика. Когда ничего никуда не едет - и джойстик в нейтральном положении.

>  чтобы поехать на зад надо перевести джойстик в другую сторону и
ему всё равно с магнитной фиксацией или с обычной всё равно включать руками.

А если надо не назад, а вверх/вниз. Ну или надо доехать до стопа, остановиться сделать некие действия, переставить стоп, доехать до следующего. Ну, там, канавки я точу хитрые, например. Можно много сценариев придумать.

**Reply by Gescha5555** [reply]
Тут я вам скажу что каждый делает под себя что ему удобно что нравится хотя это может быть 
совершенно бесполезно ,сам такой же . Сделать такой джойстик не так и сложно и напечатать на
3D принторе , пробуйте.

**Comment by andreev-** [comment]
скажите если просто к имеющимся направлениям продублировать джойстик это не то будет?

**Reply by Gescha5555** [reply]
Надо ещё перепрограммировать логику работы нажатия кнопок.

**Comment by andreev-** [comment]
![IMG_20240303_172304](https://github.com/kachurovskiy/nanoels/assets/10375078/a0d5f927-82cd-43d0-bf67-dd81c2ce7daf)

**Reply by Gescha5555** [reply]
Можно и так сделать подправив прошивку. У меня станок большой поэтому джойстик на длинном проводе ,
и подключать провод к имеющийся клавиатуре я не стал из за помех.

**Comment by andreev-** [comment]
Понял, в арограмироыании я не силен😟
--
Отправлено из Mail.ru для Android воскресенье, 03 марта 2024г., 17:40 +03:00 от Gescha5555  ***@***.*** :

**Reply by Gescha5555** [reply]
Ребята думаете я в программировании бог?
Мне 56лет и с 14 лет я занимался метала обработкой на станках и сваркой, электроника это хобби,
посидел 2 недели по разбирал код , экспериментировал и так и сяк , что  то вот сделал и знаю
что не идеально, кто то может улучшит ,только вместе можно добиться идеалла. Всё в наших руках 
пробуйте ,экспериментируйте.

**Reply by andreev-** [reply]
меня хватило только на подстановку русcких символов  :)

**Reply by Gescha5555** [reply]
Мы с вами почти одного уровня .

**Reply by Gescha5555** [reply]
А вы давно пользуетесь  ELS H4 хотелось  узнать ваше мнение о логике и удобству использования ,
а то мне кажется одному мне не удобно ей пользоваться все молчат.

**Reply by andreev-** [reply]
У меня нет другого опыта с гитарами, просто в прошлом
 токарь, пока мало этой гитарой пользовался, думаю можно привыкнуть

**Reply by andreev-** [reply]
Как с Вами связатся? тут не полуяается отправить личное сообщение

**Reply by Gescha5555** [reply]
g5555a@yandex.ru

**Comment by Gescha5555** [comment]
И так кому интересно схема
Кнопки подключены к пустым ногам ESP32
IO37- D_Left
IO36 - D_Right
IO35 - D_Up
IO45 - D_Down
IO46 - D_Speed
Все кнопки подтянуты к 3v через резисторы 5.1кО
![IMG_20240304_155228_204](https://github.com/kachurovskiy/nanoels/assets/159788822/f3e669cf-b4a1-4028-80ed-5819d534cebe)

**Reply by vegetate7** [reply]
> И так кому интересно схема
Кнопки подключены к пустым ногам ESP32

Прям к ножкам что-ли? Ну, смелый ход.
Вопрос: написано что IO35-37 подключены к PSRAM. Похоже в коде она не используется, но не мешает-ли? Или на модуле на самом деле она не распаяно?

А так идея хорошая. Можно туда MODBUS повесить, и цеплять все, что душа пожелает.

**Reply by Gescha5555** [reply]
У меня вроде работает а так пробуйте

**Comment by Gescha5555** [comment]
Теперь код , начнем с конца что бы не нарушать номера строк.
3349
 if(digitalRead(D_Left)==0 && digitalRead(D_Speed)==0){
     buttonLeftPressed = true;
  } else {buttonLeftPressed = false;}
  if(digitalRead(D_Right)==0 && digitalRead(D_Speed)==0){
   buttonRightPressed = true;
  } else {buttonRightPressed = false;}
   if(digitalRead(D_Up)==0 && digitalRead(D_Speed)==0){
     buttonUpPressed = true;
  } else {buttonUpPressed = false;}
   if(digitalRead(D_Down)==0 && digitalRead(D_Speed)==0){
     buttonDownPressed = true;
  } else { buttonDownPressed = false;}

В конце должна остаться скобка -  }
Далее идём строка 1920 и заменяем  целиком функцию
void IRAM_ATTR onAsyncTimer() {
 Axis* a = (digitalRead(D_Left)==0 || digitalRead(D_Right)==0 ) ? &z : &x;
  if (!isOn || a->movingManually || (mode != MODE_ASYNC && mode != MODE_A1)) {
    return;
   }
     else if ((digitalRead(D_Left)==0 || digitalRead(D_Up)==0) && a->pos < a->leftStop) {
       setDir(a, true);
       if (a->pos <= a->motorPos) {
       a->pos++;
       }
     a->motorPos++;
     a->posGlobal++;
   }
      else if ((digitalRead(D_Right)==0 || digitalRead(D_Down)==0) && a->pos > a->rightStop) {
       setDir(a, false);
    if (a->pos >= a->motorPos + a->backlashSteps) {
      a->pos--;
    }
    a->motorPos--;
    a->posGlobal--;
   //}
  } else {
    return;
  }
  
 DLOW(a->step);
  a->stepStartUs = micros();
  delayMicroseconds(10);
  DHIGH(a->step);
  }

1618 В setup вставляем
pinMode(D_Left, INPUT_PULLUP);
  pinMode(D_Right, INPUT_PULLUP);
  pinMode(D_Up, INPUT_PULLUP);
  pinMode(D_Down, INPUT_PULLUP);
  pinMode(D_Speed, INPUT_PULLUP); 

98 вставляем
#define D_Left 37
#define D_Right 36
#define D_Up 35
#define D_Down 45
#define D_Speed 46
Пока всё пробуйте.

**Comment by Gescha5555** [comment]
Дополнение
Замените в самом конце  то что ранее вставляли на этот. Будут работать кнопки на перемещение осей
которые на корпусе  ELS расположены.
Кто хочет попробывать джойстик  на родной конструкции от Максима без пайки - подключайте контакты 
кнопок к незадействованным осям в заде на разъёме и не забудьте прописать куда подключили
Один человек это сделал всё работает /andreev-
///////////////////////////////////////////////////////////////////
if(digitalRead(D_Left)==0){
   buttonLeftPressed = false;
   if(digitalRead(D_Speed)==0)buttonLeftPressed = true;
  }
  if(digitalRead(D_Right)==0){
   buttonRightPressed = false;
    if(digitalRead(D_Speed)==0)buttonRightPressed = true;
  } 
  if(digitalRead(D_Up)==0){
   buttonUpPressed = false;
   if(digitalRead(D_Speed)==0)buttonUpPressed = true;
  }
  if(digitalRead(D_Down)==0){
     buttonDownPressed = false;
    if(digitalRead(D_Speed)==0) buttonDownPressed = true; 
  }

**Comment by andreev-** [comment]
#define D_Left 10//37
#define D_Right 11//36
#define D_Up 12//35
#define D_Down 13//45
#define D_Speed 9//46

**Comment by andreev-** [comment]
и еще надо общий контакт с джойстика и кнопки присоединить на минус

**Comment by Gescha5555** [comment]
Перенёс обработку ускоренной подачи в функцию обработки кнопок  void taskKeypad(void *param) . У меня пропали подёргивания при ускоренном  перемещении.

**Comment by andreev-** [comment]
есть кто разбирается в коде? возможно ли прописать джойстик не в асинхронный режим  как у Gescha5555 а в основной normal mode 
1) убрать из режима  управление кнопками пуск и стоп
2) Движение осей должно осуществляться только при нажатии кнопок джойстика(Up,Down,Left,Right) 
3)   Во время движения осей регулировать подачу кнопками + или -
4 ) Кнопка Speed совместно с одной из кнопок(Up,Down,Left,Right) это ускоренное перемещение

**Comment by andreev-** [comment]
джойстик от геши5555 
для прошивки 12 версии 
с экспрессиф 2.17
5 контактов кнопочного джойстика
 подключить к A1 A2
на контакты А11,А12,А13, А21,А22.
4 направления и кнопка ускоренной
общий контакт минус
его  можно взять с энкодера


98 строка
```
#define D_Left 10
#define D_Right 11
#define D_Up 12
#define D_Down 13
#define D_Speed 9
```


1095строка
```
void taskKeypad(void *param) {
  while (emergencyStop == ESTOP_NONE) {
    processKeypadEvent();
    taskYIELD();
  
  if(digitalRead(D_Speed)==0 || digitalRead(D_Left)==0 || digitalRead(D_Right)==0 || digitalRead(D_Up)==0 || digitalRead(D_Down)==0){ 
  buttonLeftPressed = false;
  buttonRightPressed = false;
  buttonUpPressed = false;
  buttonDownPressed = false ;
 } 
  if(digitalRead(D_Left)==0 && digitalRead(D_Speed)==0){
   buttonLeftPressed = true;
  }
   if(digitalRead(D_Right)==0 && digitalRead(D_Speed)==0){
  buttonRightPressed = true;
  }
  if (digitalRead(D_Up)==0 && digitalRead(D_Speed)==0){
  buttonUpPressed = true; 
  }
  if (digitalRead(D_Down)==0 && digitalRead(D_Speed)==0){
 buttonDownPressed = true;
    }
  }
```


1619 строка
```
  pinMode(D_Left, INPUT_PULLUP);
  pinMode(D_Right, INPUT_PULLUP);
  pinMode(D_Up, INPUT_PULLUP);
  pinMode(D_Down, INPUT_PULLUP);
  pinMode(D_Speed, INPUT_PULLUP);
```


1920 строка
```
void IRAM_ATTR onAsyncTimer() {
Axis* a = (digitalRead(D_Left)==0 || digitalRead(D_Right)==0 ) ? &z : &x;
if (!isOn || a->movingManually || (mode != MODE_ASYNC && mode != MODE_A1)) {
return;
}
else if ((digitalRead(D_Left)==0 || digitalRead(D_Up)==0) && a->pos < a->leftStop) {
setDir(a, true);
if (a->pos <= a->motorPos) {
a->pos++;
}
a->motorPos++;
a->posGlobal++;
} else if ((digitalRead(D_Right)==0 || digitalRead(D_Down)==0) && a->pos > a->rightStop) {setDir(a, false);
if (a->pos >= a->motorPos + a->backlashSteps) {
a->pos--;
}
a->motorPos--;
a->posGlobal--;
} else {
  return;
}   
```



ускоренная подача работает во всех режимах ,
 просто подача во включенном асинхронном
,спасибо Генадию5555 , но хочется улучшить

**Reply by kotovasia5120** [reply]
А не лучше поставить как дополнительные кнопки на контролёр клавиатуры? У него еще достаточно не использованных каналов оставалось.

**Reply by andreev-** [reply]
наверно можно и так , но как прописать джойстик в синхронный режим что бы он работал как сейчас с этими изменениями от генадия5555  в асинхронном режиме

**Reply by kotovasia5120** [reply]
> наверно можно и так , но как прописать джойстик в синхронный режим что бы он работал как сейчас с этими изменениями от генадия5555 в асинхронном режиме

Я еще забыл учесть что клавиатура матричная и только одна кнопка одновременно обрабатывается.

**Reply by andreev-** [reply]
я развел дополнительные gpio как предложил Gescha5555 , нужен же был еще разьем на плате для джойстика
